### ISSUE TITLE
Cross extension configuration

### ISSUE THREAD

**Issue by mpacer** [jupyter/jupyter_server#118]
From a conversation in the meeting on 10/10/2019:

For classic server extensions loaded through `load_server_extension` it passed through all configuration set in traitlets down to the extensions.

The filtering of configuration based on the base application class would mean it is difficult to communicate configuration for server extension `B` when another server extension (`A`) is being used as the main `ExtensionApp`. 

We use this in `bookstore` to get access to the relevant s3 configuration parameters that are [set on the BookstoreSettings Configurable](https://github.com/nteract/bookstore/blob/master/bookstore/handlers.py#L58-L64):

```python
def load_jupyter_server_extension(nb_app):
    web_app = nb_app.web_app
    host_pattern = '.*$'

    base_url = web_app.settings['base_url']

    bookstore_settings = BookstoreSettings(parent=nb_app)
```

**Comment by mpacer** [comment]
@Zsailer

**Comment by vidartf** [comment]
Just to makes sure I understand this, there are three(?) cases of wanting to access config from another extension:
1. Extension A wanting to look at the config of extension B, where A is the `ExtensionApp`.
2. Extension A wanting to look at the config of extension B, where B is the `ExtensionApp`.
3. Extension A wanting to look at the config of extension B, where C is the `ExtensionApp`.

Is this what you are talking about, and if so, are any of these cases more relevant than the others?

> The filtering of configuration based on the base application class would mean it is difficult to communicate configuration for server extension B when another server extension (A) is being used as the main ExtensionApp.

I'm unsure where this filtering happens (I'm not that familiar with this part of the code). Would you mind sharing where this happens / which bits break things compared to notebook?

**Comment by echarles** [comment]
@mpacer I don't have the answer to your question but I find it very important and we should support as much as cases as possible and should document/demo them.

I have created a [simple example PR](https://github.com/jupyter/jupyter_server/pull/117) where there is 2 extensions and how a user can launch one or the other or both at the same time.

I have not focussed on the settings aspect, but if you have a look to the source code, you should be able IMHO to hack a bit and see how the settings can be accessed or not. If you go this road, just comment on that PR with your findings and I will inject them.

**Comment by Zsailer** [comment]
This is solved by #180. 

Extensions are now loaded in two steps:
1) `init_server_extensions` discovers and initializes each extension app before the tornado web app is created. All config is parsed+loaded, first, by the ServerApp (replacing NotebookApp), then passed down to each extension. It doesn't matter which extension starts the ServerApp, the ServerApp is the single source of truth for config and all extensions inherit the general config object. See [this comment](https://github.com/jupyter/jupyter_server/issues/55#issuecomment-614133180) for configuring multiple extensions from the CLI.
2) `load_server_extensions` calles the `_load_jupyter_server_extensions` function to append handlers to the server.

**Comment by Zsailer** [comment]
Oops, didn't mean to close this yet. 

This should be closed once #180 is merged.

**Comment by Zsailer** [comment]
Solved in #180 🚀
